beesdoo_shift/models/attendance_sheet.py

```diff
@@ -0,0 +1,634 @@
+                        "end_time": self.end_time,
+                        "is_regular": not is_compensation
+                        and is_regular_worker,
+                        "is_compensation": is_compensation
+                        and is_regular_worker,
+                    }
+                )
+            added_shift.task_id = actual_shift.id
+
+        self.validated_by = user
+        self.state = "validated"
+        return
+
+    @api.multi
+    def validate_with_checks(self):
+        self.ensure_one()
+        start_time_dt = fields.Datetime.from_string(self.start_time)
+
+        if self.state == "validated":
+            raise UserError(_("The sheet has already been validated."))
+        if start_time_dt > datetime.now():
+            raise UserError(
+                _("You must wait for the shifts to begin to validate sheet.")
+            )
+        if not self.worker_nb_feedback:
+            raise UserError(
+                _("Please give your feedback about the number of workers.")
+            )
+
+        # Fields validation
+        for added_shift in self.added_shift_ids:
+            if not added_shift.worker_id:
+                raise UserError(
+                    _("Worker name is missing for an added shift.")
+                )
+            if added_shift.state != "done":
+                raise UserError(
+                    _("Shift State is missing or wrong for %s")
+                    % added_shift.worker_id.name
+                )
+            if not added_shift.task_type_id:
+                raise UserError(
+                    _("Task Type is missing for %s")
+                    % added_shift.worker_id.name
+                )
+            if not added_shift.working_mode:
+                raise UserError(
+                    _("Working mode is missing for %s")
+                    % added_shift.worker_id.name
+                )
+
+        for expected_shift in self.expected_shift_ids:
+            if not expected_shift.state:
+                raise UserError(
+                    _("Shift State is missing for %s")
+                    % expected_shift.worker_id.name
+                )
+            if (
+                expected_shift.state == "absent"
+                and not expected_shift.compensation_no
+            ):
+                raise UserError(
+                    _("Compensation number is missing for %s")
+                    % expected_shift.worker_id.name
+                )
+
+        # Open a validation wizard only if not admin
+        if self.env.user.has_group("beesdoo_shift.group_cooperative_admin"):
+            self._validate(self.env.user.partner_id)
+            return
+        return {
+            "type": "ir.actions.act_window",
+            "res_model": "beesdoo.shift.sheet.validate",
+            "view_type": "form",
+            "view_mode": "form",
+            "target": "new",
+        }
+
+    @api.model
+    def _generate_attendance_sheet(self):
```

Comment:
Complexity of this function is not ideal, would appreciate some feedback !

Reply:
Je vois pas le problème. :/